docs/src/tutorials/write-my-software-management-plugin.md

```diff
@@ -0,0 +1,374 @@
+# Write my software management plugin
+
+**thin-edge.io** provides Software Management plugins natively for APT (Debian).
+However, there are many package management systems in the world,
+and you may want to have a plugin that is suitable for your device.
+For such a demand, we provide **Software Management Plugin API** to write a custom Software Management plugin.
+
+In this tutorial, we will look into the **Software Management Plugin API**,
+and learn how to write your own plugin with a docker plugin shell script example.
+
+## Create a plugin
+
+Create a _docker_ file in the directory _/etc/tedge/sm-plugins/_. 
+A plugin must be an executable file and located in the directory.
+
+Filename: /etc/tedge/sm-plugins/docker
+
+```shell
+#!/bin/bash
+
+set -e
+
+COMMAND="$1"
+IMAGE_NAME="$2"
+VERSION_FLAG="$3"
+IMAGE_TAG="$4"
+
+case "$COMMAND" in
+    list)
+        sudo docker image list --format '{"name":"{{.Repository}}","version":"{{.Tag}}"}'
+        ;;
+    install)
+        sudo docker pull $IMAGE_NAME:$IMAGE_TAG
+        ;;
+    remove)
+        sudo docker rmi $IMAGE_NAME:$IMAGE_TAG
+        ;;
+    prepare)
+        ;;
+    finalize)
+        ;;
+    update-list)
+        exit 1
+        ;;
+esac
+exit 0
+```
+
+> Note: the filename will be used as a plugin type to report the software list to a cloud.
+> If you name it `docker.sh`, you will see `docker.sh` as a plugin type in cloud.
+
+If you execute `./docker list`, you will see this kind of output.
+
+```
+{"name":"alpine","version":"3.14"}
+{"name":"eclipse-mosquitto","version":"2.0-openssl"}
+```
+
+The Software Management Agent runs executable plugins with a special argument, like `list`.
+Let's call the pre-defined argument such as `list`, `install`, and `remove` a **command** here. 
+As you can see from this example, a plugin should be an executable file 
+that accepts the commands and outputs to stdout and stderr in the defined JSON format. 
+Hence, you can implement a plugin in your preferred language.
```

Comment:
This is a key point. This is why it's better to say it in the introduction.

Reply:
Added it also to the beginning.

Reply:
I don't see where.

Reply:
Line 7.